### PR TITLE
Fix concurrent sort of SNI hosts and CIDR

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/28861

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -1912,12 +1912,14 @@ func buildListener(opts buildListenerOpts, trafficDirection core.TrafficDirectio
 				}
 			}
 			if !fullWildcardFound {
-				sort.Strings(chain.sniHosts)
+				chain.sniHosts = append([]string{}, chain.sniHosts...)
```
Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1915]: what effect does this have? looks like its appending a slice to an empty slice and assigning it to the original slice

Comment (adiprerepa) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1915]: what effect does this have? looks like its appending a slice to an empty slice and assigning it to the original slice

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1915]: Before: chain.sniHosts points to some point in memory storing the slice. Many other chain.sniHosts also point to it. This results in calling sort.Strings on the same array concurrently which is not legal
After: chain.sniHosts points to some point in memory storing the slice. We copy that slice into our own point in memory. Many other chain.sniHosts also have a slice, but each on is its own copy. sort.Strings is no longer called concurrently on the same slice.

Comment (adiprerepa) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1915]: ah, makes sense. thanks!

**Review by adiprerepa — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 1 above)
Comment (nirvanagit) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1915]: @howardjohn - I see the code is running in a loop, but don't see it in a go routine, so how is it concurrent?

Comment (adiprerepa) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1915]: The assumption being made is that its being modified somewhere else concurrently.

See https://github.com/istio/istio/issues/28861#issuecomment-727056641.

**Comment by hzxuzhonghu** [comment]
@howardjohn This is a great find, but i doubt if it is necessary to sort

**Comment by howardjohn** [comment]
> @howardjohn This is a great find, but i doubt if it is necessary to sort

Envoy doesn't care about the order, but we must send it deterministically (even across istiod instances). Perhaps that is possible without sort - not 100% sure though, would need to dig into the full flow better

**Comment by hzxuzhonghu** [comment]
The fields are from vs slice, worth a try later. I will approve first, and this should be backported.

**Review by hzxuzhonghu — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 1 above)
Comment (nirvanagit) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1915]: Cool, thanks! 
Thanks @howardjohn for the fix.

**Pull request merged**